[ao][sparsity] Serialization support

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* __->__ #80890
* #80889
* #80888
* #80887
* #80886

Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

Commit:
[ao][sparsity] Serialization support

Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

z-a-f commented:
I am not 100% conovonced that we can serialize simple functions. One way around it is to force a user to have "nn.Module` instances as callables for the `aggregate_fn`, etc. That way you can easily serialize them. I think that's what they do in the parametrizations.

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Serialization support"


Implemented dumping and loading of state_dicts and __get_state__ and __set_state__ functions.
hook and layer are removed from the data_groups dictionary before serializing.

In the future, might have to treat functions differently before serializing. Currently, it is being
treated similar to other types while serializing.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
